pymongo/mongo_client.py

```diff
@@ -1487,14 +1487,35 @@ def __getitem__(self, name):
         """
         return database.Database(self, name)
 
-    def _close_cursor(self, cursor_id, address):
-        """Send a kill cursors message with the given id.
+    def _cleanup_cursor(self, locks_allowed, cursor_id, address, sock_mgr,
+                        session, explicit_session):
+        """Cleanup a cursor."""
+        if locks_allowed:
+            if cursor_id:
+                if sock_mgr and sock_mgr.more_to_come:
+                    # If this is an exhaust cursor and we haven't completely
+                    # exhausted the result set we *must* close the socket
+                    # to stop the server from sending more data.
+                    sock_mgr.sock.close_socket(
+                        ConnectionClosedReason.ERROR)
+                else:
+                    self._close_cursor_now(
+                        cursor_id, address, session=session,
+                        sock_mgr=sock_mgr)
+            if sock_mgr:
+                sock_mgr.close()
+            if session and not explicit_session:
+                session._end_session(lock=True)
+        else:
+            # The cursor will be closed later in a different session.
+            if cursor_id or sock_mgr:
+                self._close_cursor(cursor_id, address, sock_mgr)
+            if session and not explicit_session:
+                session._end_session(lock=False)
 
-        What closing the cursor actually means depends on this client's
-        cursor manager. If there is none, the cursor is closed asynchronously
-        on a background thread.
-        """
-        self.__kill_cursors_queue.append((address, [cursor_id]))
+    def _close_cursor(self, cursor_id, address, sock_mgr=None):
```

Comment:
Can we rename this `_close_cursor_soon`?

Reply:
Done.